presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -729,6 +736,65 @@ public void createTable(ConnectorSession session, ConnectorTableMetadata tableMe
                 new PartitionStatistics(basicStatistics, ImmutableMap.of()));
     }
 
+    @Override
+    public ConnectorTableHandle createTemporaryTable(ConnectorSession session, List<ColumnMetadata> columns, Optional<ConnectorPartitioningMetadata> partitioning)
+    {
+        String schemaName = getTemporaryTableSchema(session);
+        String tableName = "__presto_temporary_table_" + randomUUID().toString().replaceAll("-", "_");
+        HiveStorageFormat storageFormat = getTemporaryTableStorageFormat(session);
+        Optional<HiveBucketProperty> bucketProperty = partitioning.map(partitioning1 -> {
+            HivePartitioningHandle partitioningHandle = (HivePartitioningHandle) partitioning1.getPartitioningHandle();
+            return new HiveBucketProperty(partitioning1.getPartitionColumns(), partitioningHandle.getBucketCount(), ImmutableList.of());
+        });
+        bucketProperty.ifPresent(property -> {
+            Set<String> allColumns = columns.stream()
+                    .map(ColumnMetadata::getName)
+                    .collect(toImmutableSet());
+            if (!allColumns.containsAll(property.getBucketedBy())) {
+                throw new PrestoException(INVALID_TABLE_PROPERTY, format(
+                        "Bucketing columns %s not present in schema",
+                        Sets.difference(ImmutableSet.copyOf(property.getBucketedBy()), ImmutableSet.copyOf(allColumns))));
+            }
+        });
+
+        List<HiveColumnHandle> columnHandles = getColumnHandles(columns, ImmutableSet.of(), typeTranslator);
+        storageFormat.validateColumns(columnHandles);
+
+        LocationHandle locationHandle = locationService.forNewTable(metastore, session, schemaName, tableName);
```

Comment:
That's a good  point. One question is, should "CREATE TABLE" + "INSERT INTO" in the *same* transaction be considered as having the same semantic to "CREATE TABLE AS" ? If so then using `forNewTable` also seems OK.

IIRC @arhimondr thinks it's useful during "section" level retry :)

Reply:
@shixuan-fan Current mode is already the `DIRECT_TO_TARGET_EXISTING_DIRECTORY`: https://github.com/prestodb/presto/pull/12464/commits/313d3a6753dc1f850c08d1a021cea4f5355e900d#diff-623692208e3a06e3faea192e8bc5d57bR71

However that's not the way we want to eventually go. Eventually we would like to use the `DIRECT_TO_TARGET_NEW_DIRECTORY`, so the commit is atomic. Atomic commit is needed so we can restart a stage that inserts into a temporary table.

Reply:
I chose the `DIRECT_TO_TARGET_EXISTING_DIRECTORY` method as one that is easier to implement. As the `DIRECT_TO_TARGET_NEW_DIRECTORY` requires augmented table handles for in transaction read. Not that it is a rocket science to implement, but it can be done as a separate patch.

Reply:
@wenleix 

> One question is, should "CREATE TABLE" + "INSERT INTO" in the same transaction be considered as having the same semantic to "CREATE TABLE AS" ? If so then using forNewTable also seems OK.

So currently it is not allowed to have multiple inserts into a single unpartitioned table in a single transaction. With current implementation it is not even allowed to have multiple **attempts** to insert into a temporary table. This has to be changed if we want to support stage level retries. Most likely we would have to use the `DIRECT_TO_TARGET_NEW_DIRECTORY`, so we can ensure that the commit is atomic, and no garbage will be produced by any previous unsuccessful writes.